c10/core/TensorImpl.h

```diff
@@ -463,36 +455,8 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   //
   // Note [Tensor versus Variable in C++]
   // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-  // Autograd methods are only valid for the Variable::Impl subclass
-  // of Tensor.  This is due to some questionable life choices, where
-  // a Variable has a Tensor (so they are not the same thing), but
-  // a Variable is a Tensor (they are subclassed, so that you can write
-  // code on Tensor that works both with Variables and Tensors.  Poor
-  // man's polymorphism).  Variable does NOT satisfy the Liskov Substitution
-  // Principle for Tensor; generally you want to work with all Variables,
-  // or all Tensors, but not a mix of both.  We intend to fix this in
-  // the future.
-  //
-  // Note [We regret making Variable hold a Tensor]
-  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-  // Tensor has a bunch of fields in it.  Are those fields always valid?
-  // Not necessarily: the Variable::Impl subclass of a tensor doesn't use these
-  // fields; instead, it *forwards* them to a contained, inner tensor
-  // (the 'data' tensor).  It doesn't even bother keeping the fields on the
-  // outer tensor up-to-date, because an end user could grab the inner
-  // tensor and directly, e.g., resize it (making any outer fields we track
-  // stale).
-  //
-  // As you might imagine, this is a TERRIBLE state of affairs to be in.
-  // It makes implementing everything on TensorImpl complicated: if
-  // you directly access a field on TensorImpl, you must *virtualize*
-  // the function, if you want it to work correctly when called from
-  // Variable (because we need to override the method to avoid looking
-  // in our fields, and look in the data tensor's fields.)  Anything that
-  // isn't virtualized, won't work if called on a variable.
-  //
-  // The way to fix this is to make Variable::Impl stop holding a tensor;
-  // instead, it should just *be* a tensor.
+  // Autograd methods are only valid for Variables (i.e. Tensors that contain
+  // autograd metadata).
```

Comment:
I don't think we need this comment anymore. And it should be inaccurate, because Tensors that don't contain autograd metadata can still be queried about things like `requires_grad`; it should just report false.

Reply:
I think this comment still holds true as of this PR. I will fix the comment when we have the optimization for nullifying AutogradMeta when `requires_grad=false`.